disco/voice/client.py

```diff
@@ -78,7 +90,78 @@ def send_frame(self, frame, sequence=None, timestamp=None):
 
     def run(self):
         while True:
-            self.conn.recvfrom(4096)
+            data, addr = self.conn.recvfrom(4096)
+
+            # Check the packet size
+            if len(data) < 13:
+                raise ValueError('packet is too small: {}'.format(data))
+
+            # Unpack header
+            check, seq, ts, ssrc = struct.unpack_from(self._FORMAT, data)
+            header = data[:12]
+            buff = data[12:]
+
+            # Check the packet is valid
+            if check != self._CHECK and check != self._CHECK2:
+                fmt = 'packet has invalid check bytes: {}'
+                raise ValueError(fmt.format(data))
+
+            # Decrypt data
+            nonce = bytearray(24)
+            nonce[:12] = header
+            buff = self.vc.secret_box.decrypt(bytes(buff), bytes(nonce))
+
+            if buff[0] == 0xBE and buff[1] == 0xDE:  # RFC5285 Section 4.2: One-Byte Header
+                # Please note: This has been added to future-proof the code however I have been
+                # unable to find any voice clients that are using the one-byte headers. As such,
+                # this code is untested but should work.
+                rtp_header_extension_length = buff[2] << 8 | buff[3]
+                index = 4
+                for i in range(rtp_header_extension_length):
+                    byte = buff[index]
+                    index += 1
+                    if byte == 0:
+                        continue
+
+                    l = (byte & 0b1111) + 1
+                    index += l
+
+                while buff[index] == 0:
+                    index += 1
+
+                buff = buff[index:]
+            elif check == self._CHECK2:
+                # Packets starting with b'\x90' need the first 8 bytes ignored BecauseDiscord(tm)
+                buff = buff[8:]
+
+            if ssrc not in self._decoders:
+                self._decoders[ssrc] = OpusDecoder(48000, 2)
+
+            # Lookup the SSRC and then get the user
+            user_id = 0
+            member = None
+
+            if ssrc in self.vc.ssrc_lookup:
+                user_id = int(self.vc.ssrc_lookup[ssrc])
+
+                member = self.vc.channel.guild.get_member(user_id)
+            else:
+                self.log.warning('User speaking was unknown! Dropping packet.')
+                return
+
+            buff = self._decoders[ssrc].decode(buff)
+
+            obj = VoiceReceived()
```

Comment:
Realistically we'll want this to be a named tuple, thus we get slots and avoid some of the class garbage.